lib/spack/spack/spec.py

```diff
@@ -642,14 +652,22 @@ def satisfies(self, other, strict=False):
             return all(set(self[f]) == set(other[f])
                        for f in other if (other[f] != [] and f in self))
 
+    def compatible(self, other):
+        if self.spec and self.spec._concrete:
```

Comment:
If this is checking whether `self` can be constrained by `other`, then `self` should never be concrete. As it is, this logic appears to be the same as `satisfies` (though rearranged somewhat) - does anything fail if the implementation is just:

```
return other.satisfies(self)
```

?